src/protocol/libp2p/kademlia/query/get_record.rs

```diff
@@ -145,96 +165,342 @@ impl GetRecordContext {
-        // the status is determined by whether a record was found
-        if self.pending.is_empty() && self.candidates.is_empty() {
-            match self.record_count + self.found_records.len() {
-                0 => return Some(QueryAction::QueryFailed { query: self.query }),
-                _ => return Some(QueryAction::QuerySucceeded { query: self.query }),
-            }
+        // These are the records we knew about before starting the query and
+        // the records we found along the way.
+        let known_records = self.config.record_count + self.found_records.len();
+
+        // If we cannot make progress, return the final result.
+        // A query failed when we are not able to identify one single record.
+        if self.is_done() {
+            return if known_records == 0 {
+                Some(QueryAction::QueryFailed {
+                    query: self.config.query,
+                })
+            } else {
+                Some(QueryAction::QuerySucceeded {
+                    query: self.config.query,
+                })
+            };
+        }
+
+        // Check if enough records have been found
+        let sufficient_records = self.config.sufficient_records(self.found_records.len());
+        if sufficient_records {
+            return Some(QueryAction::QuerySucceeded {
+                query: self.config.query,
+            });
+        }
+
+        // At this point, we either have pending responses or candidates to query; and we need more
+        // records. Ensure we do not exceed the parallelism factor.
+        if self.pending.len() == self.config.parallelism_factor {
+            return None;
+        }
+
+        self.schedule_next_peer()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::protocol::libp2p::kademlia::types::ConnectionType;
+
+    fn default_config() -> GetRecordConfig {
+        GetRecordConfig {
+            local_peer_id: PeerId::random(),
+            quorum: Quorum::All,
+            record_count: 0,
+            replication_factor: 20,
+            parallelism_factor: 10,
+            query: QueryId(0),
+            target: Key::new(vec![1, 2, 3].into()),
+        }
+    }
+
+    fn peer_to_kad(peer: PeerId) -> KademliaPeer {
+        KademliaPeer {
+            peer,
+            key: Key::from(peer),
+            addresses: vec![],
+            connection: ConnectionType::Connected,
         }
+    }
+
+    #[test]
+    fn config_check() {
+        // Quorum::All with no known records.
+        let config = GetRecordConfig {
+            quorum: Quorum::All,
+            record_count: 0,
+            replication_factor: 20,
+            ..default_config()
+        };
+        assert!(config.sufficient_records(20));
+        assert!(!config.sufficient_records(19));
+
+        // Quorum::All with 10 known records.
+        let config = GetRecordConfig {
+            quorum: Quorum::All,
+            record_count: 10,
```

Comment:
nit: It feels a little bit odd passing 10 where only 1 or 0 can be in real life... Also below.